### ISSUE TITLE
100% CPU usage

### ISSUE THREAD

**Issue by Patryk27** [bastion-rs/bastion#109]
Hi,

First of all: your project seems really neat and I'm on my way trying to create my first application using it :-)

Having that said:

Version: 0.3
Platform: Linux (kernel 5.3.8), Intel x86-64
Subsystem: bastion-executor
Compiler: Rust nightly (so it's `1.41.0-nightly (e87a205c2 2019-11-27)` now)

Running any of the provided examples (or just even plainly invoking `Bastion::init()`) makes my CPU usage immediately jump to 100% (on a single core); the examples themselves are working fine, it's just that I wouldn't expect such CPU load from an application that's basically idling.

`perf` told me that 99% of the time is spent inside the load balancing thread:
```rust
    pub fn sample() {
        thread::Builder::new()
            .name("load-balancer-thread".to_string())
            .spawn(move || {
                loop {
                    if let Ok(mut stats) = load_balancer::stats().try_write() {
                        // Write latest downscaled mean to statistics
                        stats.mean_level = stats
                            .smp_queues
                            .values()
                            .sum::<usize>()
                            .wrapping_div(placement::get_core_ids().unwrap().len()); // <- the `get_core_ids()` is the actual hot spot here
                    }

                    // Try sleeping for a while to wait
                    thread::sleep(Duration::new(0, 10));
                    // Yield immediately back to os so we can advance in workers
                    thread::yield_now();
                }
            })
            .expect("load-balancer couldn't start");
    }
```
... and so I've come up with two changes that reduce CPU usage to 10%[1]:

1. Since all we're doing here is `placement::get_core_ids().unwrap().len()`, I've created `placement::count_core_ids()` function that returns number of active cores:
    ```rust
    pub fn count_core_ids() -> usize {
        let mut count = 0;

        if let Some(full_set) = get_affinity_mask() {
            for i in 0..CPU_SETSIZE as usize {
                if unsafe { CPU_ISSET(i, &full_set) } {
                    count += 1;
                }
            }
        }

        count
    }
    ```
2. I've changed the `thread::sleep(Duration::new(0, 10));` into `thread::sleep(Duration::from_millis(1));`.

I'm not sure if my changes are *correct* correct, but at least they brought the CPU usage down for now.

So I guess my question says: is the high CPU usage an expected behaviour? If so, what can I do to avoid it?

Thanks,
Patryk.

[1] which, to be fair, is still kinda unsatisfactory, because all that time is "wasted" counting active cores

**Comment by Patryk27** [comment]
Hmm, other weird things happen now: having my "patch" applied, the application is *sometimes* (50% of the runs) getting stuck inside the `affine_steal()` function, consuming the 100% CPU time back again.

**Comment by vertexclique** [comment]
Hi, @Patryk27 thanks for reporting. Can you supply a minimal example that made you hit this issue?

**Comment by Patryk27** [comment]
Sure - it was virtually:
```rust
fn main() {
  Bastion::init();
  std::thread::sleep(std::time::Duration::from_millis(2000));
}
```

**Comment by vertexclique** [comment]
Now topology will be fetched only once, at the very beginning of the runtime initialization. This should decrease the sampler's usage.

**Comment by Patryk27** [comment]
Great, thanks! :-) I'll check it later, but seems like it should be working fine now.

**Comment by vertexclique** [comment]
Closing this, soon we will introduce a more advanced scheduler. Feel free to reopen it.

Small note: The main reason for CPU bump is basically runtime is looking for work to distribute to all cores with respect to core pinning.

**Comment by vertexclique** [comment]
@Patryk27 Hi Patryk! You can give it a try. This issue is resolved after #115 .

**Comment by Patryk27** [comment]
Hi, right now it works flawlessly, thanks! :heart:
